bdai_ros2_wrappers/bdai_ros2_wrappers/subscription.py

```diff
@@ -78,3 +81,106 @@ def wait_for_message(
+        node: optional node for temporary topic subscription, defaults to
+            queue_size
+        kwargs: additional arguments passed into `wait_for_messages_async`.
+
+    See `wait_for_messages_async` documentation for a reference on
+    additional keyword arguments.
+    """
+    node = node or scope.node()
+    if node is None:
+        raise ValueError("no ROS 2 node available (did you use bdai_ros2_wrapper.process.main?)")
+    future = wait_for_messages_async(topics, mtypes, node=node, **kwargs)
+    if not wait_for_future(future, timeout_sec, context=node.context):
+        future.cancel()
+        return None
+    return future.result()
+
+
+def wait_for_messages_async(
+    topics: typing.List,
+    mtypes: typing.List,
+    *,
+    queue_size: int = 10,
+    delay: float = 0.2,
+    allow_headerless: bool = False,
+    sleep: float = 0.5,
+    qos_profiles: typing.Optional[typing.Dict[str, rclpy.qos.QoSProfile]] = None,
+    node: typing.Optional[rclpy.node.Node] = None,
+    callback_group: typing.Optional[rclpy.callback_groups.CallbackGroup] = None,
+) -> rclpy.task.Future:
+    """Asynchronous version of wait_for_messages
+
+    Args:
+        topics: List of topics
+        mtypes: List of message types, one for each topic.
+        queue_size: synchronizer message queue size
+        delay: The delay in seconds for which the messages
+            could be synchronized (i.e. the time window).
+        allow_headerless: Whether it's ok for there to be
+            no header in the messages.
+        sleep: the amount of time to wait before checking
+            whether messages are received
+        qos_profiles: maps from topic name to QoSProfile
+        node: optional node for temporary topic subscription, defaults to
+            the current process-wide node (if any).
+        callback_group: callback group for the message filter subscribers
+    """
+    node = node or scope.node()
+    if node is None:
+        raise ValueError("no ROS 2 node available (did you use bdai_ros2_wrapper.process.main?)")
+    if qos_profiles is None:
+        qos_profiles = {}
+    subs: typing.List[message_filters.Subscriber] = []
+    for topic, mtype in zip(topics, mtypes):
+        qos_profile = qos_profiles.get(topic, 10)
+        subs.append(
+            message_filters.Subscriber(node, mtype, topic, qos_profile=qos_profile, callback_group=callback_group),
+        )
+    future = rclpy.task.Future()
+
+    def callback(*messages: typing.List[MessageT]) -> None:
+        if not future.done():
+            future.set_result(messages)
+
+    ts = message_filters.ApproximateTimeSynchronizer(
+        subs,
+        queue_size,
+        delay,
+        allow_headerless=allow_headerless,
+    )
+    ts.registerCallback(callback)
+    future.add_done_callback(lambda future: _destroy_subs(node, subs))
+    return future
+
+
+def _destroy_subs(node: rclpy.node.Node, subs: typing.List[message_filters.Subscriber]) -> None:
```

Comment:
@kaiyu-zheng FYI: you can define this done callback in `wait_for_messages_async` scope like you did for `callback`.